VisualRiders.PointOfSale.Project/Services/DiscountsService.cs

```diff
@@ -0,0 +1,195 @@
+        _discountItemsRepository = discountItemsRepository;
+        _productsRepository = productsRepository;
+        _servicesRepository = servicesRepository;
+        _mapper = mapper;
+    }
+
+    public Discount Create(CreateUpdateDiscountDto dto)
+    {
+        var discount = _mapper.Map<Discount>(dto);
+
+        discount.BusinessEntityId = 1;
+        discount.DiscountItems = new List<DiscountItem>();
+        
+        _repository.Add(discount);
+        _repository.SaveChanges();
+
+        return discount;
+    }
+
+    public List<ReadDiscountDto> GetAll()
+    {
+        return _repository.GetAll().Select(_mapper.Map<ReadDiscountDto>).ToList();
+    }
+
+    public ReadDiscountDto? GetById(int id)
+    {
+        return _mapper.Map<ReadDiscountDto>(_repository.GetById(id));
+    }
+
+    public ReadDiscountDto? UpdateById(int id, CreateUpdateDiscountDto dto)
+    {
+        var discount = _repository.GetById(id);
+
+        if (discount == null) return null;
+
+        discount = _mapper.Map<Discount>(dto); 
+
+        _repository.SaveChanges();
+
+        return _mapper.Map<ReadDiscountDto>(discount);
+    }
+
+    public ReadDiscountItemDto? AddItemById(int id, CreateUpdateDiscountItemDto dto)
+    {
+        var discountItem = _mapper.Map<DiscountItem>(dto);
+
+        var discount = _repository.GetById(id);
+
+        if (discount == null) return null;
+
+        discountItem.Discount = discount;
+
+
+        if (dto.ProductId != null)
+        {
+            var product = _productsRepository.GetById(dto.ProductId.Value);
+            if (product == null)
+            {
+                throw new UnprocessableEntity($"Product with Id = {dto.ProductId.Value} does not exist");
+            }
+
+            discountItem.Product = product;
+        }
+
+        if (dto.ServiceId != null)
+        {
+            var service = _servicesRepository.GetById(dto.ServiceId.Value);
+            if (service == null)
+            {
+                throw new UnprocessableEntity($"Product with Id = {dto.ServiceId.Value} does not exist");
+            }
+
+            discountItem.Service = service;
+        }
+
+        if(discount.DiscountItems == null)
+        {
+            discount.DiscountItems = new List<DiscountItem>();
+        }
+        discount.DiscountItems.Add(discountItem);
+
+        _discountItemsRepository.Add(discountItem);
+        _discountItemsRepository.SaveChanges();
+
+        _repository.SaveChanges();
+
+        return _mapper.Map<ReadDiscountItemDto>(discountItem);
+    }
+
+    public ReadDiscountItemDto? UpdateItemById(int discountId, int itemId, CreateUpdateDiscountItemDto dto)
+    {
+        var discount = _repository.GetById(discountId);
+
+        if (discount == null) return null;
+
+        var discountItem = _discountItemsRepository.GetById(itemId);
+
+        if (discountItem == null)
+        {
+            throw new UnprocessableEntity($"Discount item with Id = {itemId} does not exist");
```

Comment:
It is not actually right to return 422 here but as we don't have a separate exception for Not Found errors, there is no other way to implement this.